src/panels/ExplorerPanel.tsx

```diff
@@ -0,0 +1,157 @@
+            onClick={(file) => activateFileInPane({ pane: left.id, path: file.path })}
+          />
+        )}
+        {hasRightPanelFiles && <div className="ml-4 text-xs font-bold">RIGHT</div>}
+        {hasRightPanelFiles && (
+          <FileTree
+            files={right.files}
+            rightAction={(file) => (
+              <VscSplitHorizontal
+                title="Move to LEFT split pane"
+                onClick={(evt) => {
+                  evt.preventDefault();
+                  evt.stopPropagation();
+                  splitFileToPane({ file, fromPane: right.id, toPane: 'LEFT' });
+                }}
+              />
+            )}
+            root
+            selectedFiles={right.active ? [right.active] : []}
+            onClick={(file) => activateFileInPane({ pane: right.id, path: file.path })}
+          />
+        )}
+      </PanelSection>
+      <PanelSection title="Project X">
+        <FileTree
+          files={allFiles}
+          root
+          selectedFiles={[left.active, right.active].filter((file) => file).map((file) => file!)}
+          onClick={(file) => openFileInPane({ pane: DEFAULT_PANE, file })}
+        />
+      </PanelSection>
+    </PanelWrapper>
+  );
+}
+
+interface FileTreeBaseProps {
+  root?: boolean;
+  files: FileEntry[];
+  file?: FileEntry;
+  selectedFiles: FileEntry[];
+  onClick: (file: FileEntry) => void;
+  rightAction?: (file: FileEntry) => React.ReactNode;
+}
+
+const FileTree = ({ files, root, ...props }: FileTreeBaseProps) => (
+  <div className="overflow-scroll">
+    <ul
+      className={cx('', {
+        'ml-4': root,
+        'ml-6': !root,
+      })}
+    >
+      {files.map((file) => (
+        <FileTreeNode file={file} files={files} key={file.path} {...props} />
+      ))}
+    </ul>
+  </div>
+);
+
+const FileTreeNode = ({ file, onClick, rightAction, selectedFiles }: FileTreeBaseProps) => {
+  if (!file) {
+    return null;
+  }
+  const isFolder = Array.isArray(file.children);
+
+  // Hide DOT_FILES
+  if (file.name?.startsWith('.')) {
+    return null;
+  }
+
+  return (
+    <li>
+      <div
+        className={cx(
+          'mb-1 py-1',
+          'flex flex-row items-center gap-1', //
+          'group',
+          {
+            'bg-slate-100': selectedFiles.some((f) => f.path === file.path),
+            'cursor-pointer hover:bg-slate-100': !isFolder,
+          },
+        )}
+        title={file.name}
+        onClick={() => !isFolder && onClick(file)}
+      >
+        {isFolder ? <VscFolder className="shrink-0" /> : <VscFile className="shrink-0" />}
+        <span
+          className={cx('flex-1 truncate', {
+            'font-bold': isFolder,
+          })}
+        >
+          {file.name}
+        </span>
+        {rightAction && (
+          <div className="mr-2 hidden p-0.5 hover:bg-gray-200 group-hover:block">{rightAction(file)}</div>
+        )}
+      </div>
+
+      {isFolder && Array.isArray(file.children) && (
```

Comment:
@danvk somehow we need this additional check otherwise TS don't understand that `file.children` is an array! 
I thought TS would "read through variables" for this kind of type narrowing but it's not working here. 
Maybe you can help with some guidance.

Reply:
Created issue https://github.com/refstudio/refstudio/issues/75 related with this.